fastlane_bot/helpers/txhelpers.py

```diff
@@ -133,579 +61,145 @@ def validate_and_submit_transaction(
+            if allowance == 0:
+                tx = self._create_transaction(token_contract, "approve", [self.arb_contract.address, MAX_UINT256], 0)
+                self._update_transaction(tx)
+                raw_tx = self._sign_transaction(tx)
+                tx_hash = self._send_regular_transaction(raw_tx)
+                self._wait_for_transaction_receipt(tx_hash)
+
+    def _create_transaction(self, contract, fn_name: str, args: list, value: int) -> dict:
+        return {
+            "type": 2,
+            "value": value,
+            "chainId": self.chain_id,
+            "from": self.wallet_address,
+            "to": contract.address,
+            "data": contract.encode_abi(fn_name=fn_name, args=args),
+            "nonce": self.cfg.w3.eth.get_transaction_count(self.wallet_address)
+        }
 
-    @staticmethod
-    def _get_payload(method: str, params: [] = None) -> Dict:
-        """
-        Generates the request payload for the API call. If the method is "eth_estimateGas", it attaches the params
-        :param method: the API method to call
-        """
+    def _update_transaction(self, tx: dict):
+        tx["gas"] = self.cfg.w3.eth.estimate_gas(tx) # occasionally throws an exception
+        if self.use_access_list:
+            result = self.cfg.w3.eth.create_access_list(tx) # rarely throws an exception
+            if tx["gas"] > result["gasUsed"]:
+                tx["gas"] = result["gasUsed"]
+                tx["accessList"] = [
+                    {
+                        "address": access_item["address"],
+                        "storageKeys": [storage_key.hex() for storage_key in access_item["storageKeys"]]
+                    }
+                    for access_item in result["accessList"]
+                ]
+        tx.update(self.cfg.network.gas_strategy(self.cfg.w3))
 
-        if method == "eth_estimateGas" or method == "eth_sendPrivateTransaction":
-            return {"id": 1, "jsonrpc": "2.0", "method": method, "params": params}
-        else:
-            return {"id": 1, "jsonrpc": "2.0", "method": method}
+    def _sign_transaction(self, tx: dict) -> str:
+        return self.cfg.w3.eth.account.sign_transaction(tx, self.cfg.ETH_PRIVATE_KEY_BE_CAREFUL).rawTransaction
 
-    def _query_alchemy_api_gas_methods(self, method: str, params: list = None):
-        """
-        This queries the Alchemy API for a gas-related call which returns a Hex String.
-        The Hex String can be decoded by casting it as an int like so: int(hex_str, 16)
+    def _send_regular_transaction(self, raw_tx: str) -> str:
+        return self.cfg.w3.eth.send_raw_transaction(raw_tx).hex()
 
-        :param method: the API method to call
-        """
-        response = requests.post(
-            self.alchemy_api_url,
-            json=self._get_payload(method=method, params=params),
-            headers=self._get_headers,
+    def _send_private_transaction(self, raw_tx: str) -> str:
```

Comment:
Couple points:

1. This must be tested & confirmed that it works. It may be difficult to find a profitable transaction on Ethereum, so a good testing option would be to force a transaction through that is not profitable after gas costs. 
2. This function likely only works with Alchemy. For someone using a different RPC or local node it would fail unless the RPC implements the `eth_sendPrivateTransaction` method - I'm not aware of non-Alchemy RPCs that have this aside from Flashbots themselves. I still believe implementing this call directly to Flashbots would be optimal as it would remove all reliance on Alchemy.

Reply:
1. Yes - that has crossed my mind, with the main point in it being to make sure that the private-transaction mechanism is working correctly.
2. This is currently the case anyway, as far as I remember (the bot uses private transactions for Ethereum, regardless of which node is in use).